calc-example/calc/src/ir.rs

```diff
@@ -90,19 +104,169 @@ pub struct Function {
+            _ => None,
+        })
+        .next()
+}
+
+impl Program {
+    pub(crate) fn find_function(self, db: &dyn crate::Db, name: FunctionId) -> Option<Function> {
+        find_function(db, self, name)
+    }
+}
+
+// ANCHOR: anchor
+/// Anchors mark significant points in the program text,
+/// such as the start of the input or the start of a function.
+/// All other spans are relative to an anchor, ensuring that
+/// the spans don't change even when other content is added before
+/// the anchor point.
+pub trait Anchor {
+    fn anchor_location(&self, db: &dyn crate::Db) -> Location;
+}
+// ANCHOR_END: anchor
+
+// ANCHOR: anchor_impls
+impl Anchor for Program {
+    fn anchor_location(&self, _db: &dyn crate::Db) -> Location {
+        Location::start()
+    }
+}
+
+impl Anchor for Function {
+    fn anchor_location(&self, db: &dyn crate::Db) -> Location {
+        Function::anchor_location(*self, db)
+    }
+}
+// ANCHOR_END: anchor_impls
+
+/// Represents a specific location into the source string
+/// as a utf-8 offset.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Location(usize);
+
+impl Location {
+    pub fn as_usize(self) -> usize {
+        self.0
+    }
+
+    pub fn start() -> Self {
+        Self(0)
+    }
+}
+
+impl std::ops::Add<Offset> for Location {
+    type Output = Location;
+
+    fn add(self, rhs: Offset) -> Self::Output {
+        Location(self.0 + rhs.0)
+    }
+}
+
+impl std::ops::Add<usize> for Location {
+    type Output = Location;
+
+    fn add(self, rhs: usize) -> Self::Output {
+        Location(self.0 + rhs)
+    }
+}
+
+impl std::ops::AddAssign<usize> for Location {
+    fn add_assign(&mut self, rhs: usize) {
+        *self = *self + rhs
+    }
+}
+
+impl std::ops::Sub<Location> for Location {
+    type Output = Offset;
+
+    fn sub(self, rhs: Location) -> Self::Output {
+        Offset(self.0 - rhs.0)
+    }
+}
+
+/// Represents an offset in the source program relative to some anchor.
+#[derive(Copy, Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Offset(usize);
+
+impl Offset {
+    pub fn location(self, anchor: Location) -> Location {
+        Location(self.0 + anchor.0)
+    }
+}
+
+// ANCHOR: span
+/// Stores the location of a piece of IR within the source text.
+/// Spans are not stored as absolute values but rather relative to some enclosing anchor
+/// (some struct that implements the `Anchor` trait).
+/// This way, although the location of the anchor may change, the spans themselves rarely do.
+/// So long as a function doesn't convert the span into its absolute form,
+/// and thus read the anchor's precise location, it won't need to re-execute, even if the anchor
+/// has moved about in the file.
+///
+/// **NB:** It is your job, when converting the span into relative positions,
+/// to supply the correct anchor! For example, the anchor for the expressions
+/// within a function body is the function itself.
```

Comment:
This feels really awkward and I'd hope we could support something better. I wonder if something like this might work:
```rust
#[tracked]
struct Anchor {
    #[id]
    name: String,
    parent: Option<Anchor>,
    offset: Offset,
}

struct Span {
    anchor: Anchor,
    start: Offset,
    end: Offset,
}
```

Reply:
I considered that setup. 

There are a few problems that would lead me to avoid it in a real compiler:

* The `Span` is a very common data structure and it now occupies an additional 32 bits, at least. Ideally you'd have spans be super tiny so you can have them all over the place (even the current setup, with two offsets, is kinda large).
* The anchor winds up duplicating parts of the things it anchors (e.g., the name) which makes the code a bit more awkward.

The latter one is not so bad, but the former pushed me over the edge and made me happier with this setup. 

Thoughts?

Reply:
You could always intern them, but I can see that being unacceptable slow for a real compiler. I've been unable to come up with a nicer solution so I guess I'm happiest with this setup.